staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go

```diff
@@ -0,0 +1,153 @@
+         "VirtletCPUModel":"host-model"
+      },
+      "name":"vmdefault",
+      "namespace":"kube-system",
+      "tenant":"system"
+   },
+   "spec":{
+      "virtualMachine":{
+         "image":"%s",
+         "imagePullPolicy":"IfNotPresent",
+         "keyPairName":"foobar",
+         "name":"vm",
+         "publicKey":"ssh-rsa AAA",
+         "resources":{
+            "limits":{
+               "cpu":"%d",
+               "memory":"%dGi"
+            },
+            "requests":{
+               "cpu":"%d",
+               "memory":"%dGi"
+            }
+         }
+      }
+   }
+}
+`
+)
+
+type Network struct {
+	Uuid     string
+	Port     string
+	Fixed_ip string
+}
+
+type MetadataType struct {
+	key   string
+	value string
+}
+
+type SecurityGroup struct {
+	name string
+}
+
+type LinkType struct {
+	Link string
+	Rel  string
+}
+
+// VM creation request in Openstack
+type OpenstackRequest struct {
+	Name            string
+	Flavor          string
+	Networks        []Network
+	Key_name        string
+	Metadata        MetadataType
+	Security_groups []SecurityGroup
+	User_data       string
+}
+
+// VM creation response in Openstack
+type OpenstackResponse struct {
+	id              string
+	Links           []LinkType
+	Security_groups []SecurityGroup
+}
+
+// Convert Openstack request to kubernetes pod request body
+func ConvertToOpenstackRequest(body []byte) []byte {
+	cpu, mem := GetRequestedResource("m1.timy")
+	imageUrl := GetVmImageUrl("ImageRefID")
+
+	podJson := fmt.Sprintf(POD_JSON_STRING_TEMPLATE, imageUrl, cpu, mem, cpu, mem)
+	return []byte(podJson)
+}
+
+// Convert kubernetes pod response to Openstack response body
+func ConvertToOpenstackResponse(body []byte) []byte {
+	return body
+}
+
+func IsOpenstackRequest(req *http.Request) bool {
+	return req.Header.Get("openstack") == "true"
+}
+
+// TODO: add config map to cache a few OS images to simulate Openstack image resitry
+func GetVmImageUrl(imageRef string) string {
+	return "download.cirros-cloud.net/0.5.1/cirros-0.5.1-x86_64-disk.img"
+}
+
+// get the requested resource, CPU, RAM, from the Openstack flavor of VM
+// TODO: add config map for most used Openstack flavors
+//       consider to use resource structure instead of returning individual resources
+func GetRequestedResource(flavor string) (int, int) {
+	return 1, 2
+}
+
+// Get the tenant for the request from the request Token
+func GetTenantFromRequest(r *http.Request) string {
+	return "system"
+}
+
+// Get the namespace, maps to the Openstack projct, from the Openstack token
+func GetNamespaceFromRequest(r *http.Request) string {
+	return "kube-system"
+}
+
+// read request will void the http requet body to will need to clone here
+func DumpRequest(r *http.Request) {
+	klog.V(9).Infof("debug: dump request")
+
+	klog.V(9).Infof("Header: %v", r.Header)
+
+	r2 := r.Clone(r.Context())
+	body, err := ioutil.ReadAll(r2.Body)
+	if err != nil {
+		klog.Infof("debug: error read request body: %v", err)
```

Comment:
I am confused, should this be klog.Errorf?

Reply:
this is a debugging helper function and intended to use infof. it has done its job as it has passed the request debugging part.  i will remove the function to avoid this churn.